### PR TITLE
Julia 1.0 Compatibility

### PR DESCRIPTION
**Pull request by dcjones**
This was an ordeal, and ended up being a huge patch, but should bring sequences up to date with julia 1.0 changes.

Needs a couple other PRs to work:
  * https://github.com/BioJulia/BioSymbols.jl/pull/30
  * https://github.com/BioJulia/BioCore.jl/pull/18

### CODE REVIEW

**Comment by BenJWard** [comment]
Thanks for taking the time to do this @dcjones! I'll merge when the other two PRs are in :+1:

File: REQUIRE [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
+BioCore 2.0.0
+BioSymbols 3.0.1
 BufferedStreams
 Combinatorics
 IndexableBitVectors
-IterTools 0.1
 IntervalTrees
 Twiddle
 DataStructures
+Nullables
```
Comment (bicycle1885) [inline REQUIRE]: It is okay to have this now but I think that in general we should not depend on deprecated/removed features. We may need a second round to review our APIs.

**Comment by BenJWard** [comment]
@dcjones To take this furthur and get CI passes I'm currently waiting on YAML, the version 1.0 changes PR is waiting on a Codecs release, which is here: https://github.com/JuliaLang/METADATA.jl/pull/16489, but as it's not on attobot it's not moving very far. Can you head to the PR and usher them along a bit, or turn on attobot for Codecs and YAML? :)

**Comment by BenJWard** [comment]
@dcjones Ok I've made a series of commits trying to get tests to pass. I've got them passing for 1.0, but curously not for 0.7. What's making me tear my hair out is julia complains "ERROR: LoadError: LoadError: LoadError: syntax: invalid variable expression in "where"". Whilst pointing to "in expression starting at /home/travis/build/BioJulia/BioSequences.jl/src/bioseq/constructors.jl:65". At which is a constructor, which does not make use of the "where" expression anywhere in it's definition. What do you think?

**Comment by jgreener64** [comment]
So I had a go and could reproduce your issue @BenJWard. A MWE without imports that works on v0.7 but not on v1.0 is:

```julia
abstract type Alphabet end
abstract type Sequence end

struct DNAAlphabet{n} <: Alphabet end

mutable struct BioSequence{A<:Alphabet} <: Sequence
    data::Vector{UInt64}  # encoded character sequence data
    part::UnitRange{Int}  # interval within `data` defining the (sub)sequence
    shared::Bool          # true if and only if `data` is shared between sequences

    function BioSequence{A}(data::Vector{UInt64},
                            part::UnitRange{Int},
                            shared::Bool) where A
        return new(data, part, shared)
    end
end

function BioSequence{DNAAlphabet{4}}(seq::BioSequence{DNAAlphabet{2}}) # errors here
    newseq = BioSequence{vim}(length(seq))
    #for (i, x) in enumerate(seq)
    #    unsafe_setindex!(newseq, x, i)
    #end
    return newseq
end
```

It seems that if you replace the offending line with

```julia
function BioSequence{T}(seq::BioSequence{DNAAlphabet{2}}) where {T<:DNAAlphabet{4}}
```

then it seems to work, which is strange. I made a PR to implement this across this file and tests seem to pass on both v0.7 and v1.0.

Feel free to use it if you like, but I'm wondering if this is a Julia issue?

**Comment by BenJWard** [comment]
Thanks @jgreener64! You're option made me think of another alternative, which also passes both of 0.7 and 1.0!

```julia
abstract type Alphabet end
abstract type Sequence end

struct DNAAlphabet{n} <: Alphabet end

mutable struct BioSequence{A<:Alphabet} <: Sequence
    data::Vector{UInt64}  # encoded character sequence data
    part::UnitRange{Int}  # interval within `data` defining the (sub)sequence
    shared::Bool          # true if and only if `data` is shared between sequences

    function BioSequence{A}(data::Vector{UInt64},
                            part::UnitRange{Int},
                            shared::Bool) where A
        return new(data, part, shared)
    end
end

function (::Type{BioSequence{DNAAlphabet{4}}})(seq::BioSequence{DNAAlphabet{2}})
    newseq = BioSequence{DNAAlphabet{4}}(length(seq))
    for (i, x) in enumerate(seq)
        unsafe_setindex!(newseq, x, i)
    end
    return newseq
end
```

**Comment by BenJWard** [comment]
Merging into the `version-1.0` branch now. Release to follow shortly.

**Pull request merged**